### PR TITLE
CFINSPEC-95: Enhance `x509_certificate` resource

### PR DESCRIPTION
**Pull request by ahasunos**
✅ Signed-off-by: Sonu Saha [sonu.saha@progress.com](mailto:sonu.saha@progress.com)

## Related Issue
**CFINSPEC-95: Enhance `x509_certificate` resource**

## Description
- This PR introduces the enhancement to the existing `x509_certificate` resource with new matcher & properties.
- Using the `x509_certificate` resource
Given that the user has called the `x509_certificate` resource
then the resource allows to test with new matchers & properties.
- New matcher: `have_purpose`
- New properties: `email`, `keylength`, `subject_alt_names`
- Syntax:
  ```ruby
    describe x509_certificate(filepath: '/home/alice.crt') do
      its('email') { should eq 'alice@alice.com' }
      its('keylength') { should be >= 1024 }
      its('subject_alt_names') { should include 'DNS:example.com' }
      it { should have_purpose("SSL server CA : Yes") }
  end
  ```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: lib/inspec/resources/x509_certificate.rb [hunk 1]
```diff
@@ -138,6 +145,49 @@ def extensions
       @extensions
     end
 
+    # check purpose of the certificate
+    def has_purpose?(purpose)
```
Comment (ahasunos) [inline lib/inspec/resources/x509_certificate.rb:149]: Please check if this could be optimised and handled in a different way.

Comment (ahasunos) [reply lib/inspec/resources/x509_certificate.rb:149]: Suggestion/feedback would be helpful.

File: lib/inspec/resources/x509_certificate.rb [hunk 2]
```diff
@@ -138,6 +145,49 @@ def extensions
...
+      if @opts[:filepath]
+        cert_purpose = fetch_purpose(@opts[:filepath])
+      else
+        f = File.open("temporary_certificate.pem", "w")
+        f.write(@cert.to_pem)
+        f.rewind
+        begin
+          f = File.open("temporary_certificate.pem", "w")
+          f.write(@cert.to_pem)
+          f.rewind
```
Comment (clintoncwolfe) [inline lib/inspec/resources/x509_certificate.rb]: I don't understand why you are writing to the file twice with the same contents.

**Review by clintoncwolfe — COMMENTED** [review]
I don't understand the implementation of have_purpose - I think you need to explain it to me.

File: lib/inspec/resources/x509_certificate.rb (hunk 2 above)
Comment (ahasunos) [reply lib/inspec/resources/x509_certificate.rb]: Yes, it's redundant.

Comment (ahasunos) [reply lib/inspec/resources/x509_certificate.rb]: @clintoncwolfe 
The command to find the certificate purpose is: `openssl x509 -noout -purpose -in [the-certificate-or-the-certificate-path]`. The `-in` argument expects a certificate file or path to a certificate file, and not content of the certificate.

There are two ways InSpec initialises a x509 certificate.
1. content
2. filepath

If the certificate was initialised with filepath, I'm using the file path directly in the command `openssl x509 -noout -purpose -in [the-certificate-or-the-certificate-path]` to fetch the certificate purpose.

If the certificate was initialised with content, I'm writing the content to a temporary file and then fetching the certificate purpose.

**Review by clintoncwolfe — APPROVED** [review]

File: docs-chef-io/content/inspec/resources/x509_certificate.md [hunk 3]
```diff
@@ -110,6 +110,14 @@ More bits increase security, but at the cost of speed and in extreme cases, comp
       its('key_length') { should be 2048 }
```
Comment (dkumaras) [inline docs-chef-io/content/inspec/resources/x509_certificate.md:103]: @ahasunos - Please check this rewritten sentence.

File: docs-chef-io/content/inspec/resources/x509_certificate.md [hunk 4]
```diff
@@ -110,6 +110,14 @@ More bits increase security, but at the cost of speed and in extreme cases, comp
...
     end
 
+### keylength
+
```
Comment (dkumaras) [inline docs-chef-io/content/inspec/resources/x509_certificate.md:107]: In general, most of the properties has type. Shall we include for all?

Comment (dkumaras) [reply docs-chef-io/content/inspec/resources/x509_certificate.md:107]: @ahasunos - Pls check this too!

File: docs-chef-io/content/inspec/resources/x509_certificate.md [hunk 5]
```diff
@@ -176,3 +184,52 @@ The `extensions` hash property is mainly used to determine what the certificate
...
+    describe x509_certificate('/etc/pki/www.mywebsite.com.pem') do
+      it { should be_certificate }
+    end
+
+### have_purpose
+
+The `have_purpose` matcher tests if the certificate has the specified purpose enabled or not.
+
+    describe x509_certificate('/etc/pki/www.mywebsite.com.pem') do
```
Comment (dkumaras) [inline docs-chef-io/content/inspec/resources/x509_certificate.md:222]: Didnt get it. Will connect with you to understand.

**Review by dkumaras — COMMENTED** [review]
@ahasunos - Pls check the corrections made and confirm.

**Comment by ahasunos** [comment]
> @ahasunos - Pls check the corrections made and confirm.

Looks good 💯  @dkumaras

**Review by dkumaras — APPROVED** [review]

**Pull request merged**